pkg/dpm/manager.go

```diff
@@ -3,99 +3,208 @@ package dpm
+				dpm.shutDownPlugins(pluginsMap)
+				break HandleSignals
+			}
+		}
+	}
+}
+
+func (dpm *Manager) handleNewPlugins(currentPluginsMap map[string]devicePlugin, newPluginsList ResourceLastNamesList) {
+	var wg sync.WaitGroup
+
+	newPluginsSet := make(map[string]bool)
+	for _, newPluginLastName := range newPluginsList {
+		newPluginsSet[newPluginLastName] = true
 	}
 
-	go dpm.handleSignals()
+	// Add new plugins
+	for newPluginLastName, _ := range newPluginsSet {
+		wg.Add(1)
+		go func() {
+			if _, ok := currentPluginsMap[newPluginLastName]; !ok {
+				glog.V(3).Infof("Adding a new plugin \"%s\"", newPluginLastName)
+				plugin := newDevicePlugin(dpm.lister.GetResourceNamespace(), newPluginLastName, dpm.lister.NewPlugin(newPluginLastName))
+				startUpPlugin(newPluginLastName, plugin)
+				currentPluginsMap[newPluginLastName] = plugin
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	// Remove old plugins
+	for pluginLastName, plugin := range currentPluginsMap {
+		wg.Add(1)
+		go func() {
+			if _, ok := newPluginsSet[pluginLastName]; !ok {
+				shutDownPlugin(pluginLastName, plugin)
+				delete(currentPluginsMap, pluginLastName)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+}
 
-	// We can now move to functionality: first, the initial pool of devices is needed.
-	devices := lister.Discover()
+func (dpm *Manager) startPluginServers(pluginsMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
 
-	// As we use the pool to initialize device plugins (the actual gRPC servers) themselves.
-	for deviceClass, deviceIDs := range *devices {
-		dpm.plugins = append(dpm.plugins, lister.NewDevicePlugin(deviceClass, deviceIDs))
+	for pluginLastName, plugin := range pluginsMap {
+		wg.Add(1)
+		go func() {
+			startPluginServer(pluginLastName, plugin)
+			wg.Done()
+		}()
 	}
+	wg.Wait()
+}
 
-	return dpm
+func (dpm *Manager) stopPluginServers(pluginsMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
+
+	for pluginLastName, plugin := range pluginsMap {
+		wg.Add(1)
+		go func() {
+			stopPluginServer(pluginLastName, plugin)
+		}()
+	}
+	wg.Wait()
 }
 
-// Run starts the DevicePluginManager.
-func (dpm *DevicePluginManager) Run() {
-	defer dpm.fsWatcher.Close()
-	dpm.startPlugins()
+func (dpm *Manager) shutDownPlugins(pluginsMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
 
-	// The manager cannot be restarted, it is expected to be the `core` of an application.
-	<-dpm.stopCh
-	dpm.stopPlugins()
+	for pluginLastName, plugin := range pluginsMap {
+		wg.Add(1)
+		go func() {
+			shutDownPlugin(pluginLastName, plugin)
+			delete(pluginsMap, pluginLastName)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
 }
 
-// startPlugins is helper function to start the underlying gRPC device plugins.
-func (dpm *DevicePluginManager) startPlugins() {
-	for _, plugin := range dpm.plugins {
-		go plugin.Start()
+func startUpPlugin(pluginLastName string, plugin devicePlugin) {
+	if devicePluginImplementation, ok := plugin.DevicePlugin.(PluginInterfaceStart); ok {
```

Comment:
devicePluginImpl is slightly more consistent in what we'd normally see in go.

Reply:
Done